### ISSUE TITLE
Google Cloud Function - Error: Could not load the default credentials.

### ISSUE THREAD

**Issue by timhj** [googleapis/google-auth-library-nodejs#798]
This has only recently started happening for me. Some cloud functions which are using the Cloud Vision API have started failing due to an Auth error. The failures seem random, with requests working sometimes and other times not. As there is no explicit Auth happening (it's the Node JS GCF runtime for an existing project, it's not clear what could be the issue).

Errors look like this:
```
A 2019-10-01T03:17:00.907Z PROD-PDF-XXXX 718747420330136 
    Unhandled rejection PROD-PDF-XXXX 718747420330136
E 2019-10-01T03:17:00.910Z PROD-PDF-XXXX 718747420330136 
    Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
    at GoogleAuth.getApplicationDefaultAsync (/srv/functions/node_modules/google-auth-library/build/src/auth/googleauth.js:161:19)
    at process._tickCallback (internal/process/next_tick.js:68:7) PROD-PDF-XXXX 718747420330136 
```

Triggering code is: 
```
if(statusUpdateResult.affectedRows < 1){
        // to avoid race conditions from concurrent competing threads, lets make sure this thread 
        // was the one that updated the status first
        let concurrentThreadError = Error('Not the first to update status to processing so exiting');
        console.error(concurrentThreadError);
        return concurrentThreadError;
      } else {
        // status updated and ready for processing
        console.log(`Updated pdf id ${pdfRecord.pdf_id} to queuing_vision.`);
      }
      
      // queue the file for processing by vision API
      const gcsSourceUri = `gs://${bucketName}/${fileName}`;
      const gcsDestinationUri = `gs://${ocrJSONBucket}/${fileName}.json`;

      const inputConfig = {
        // Supported mime_types are: 'application/pdf' and 'image/tiff'
        mimeType: 'application/pdf',
        gcsSource: {
          uri: gcsSourceUri,
        },
      };
      const outputConfig = {
        gcsDestination: {
          uri: gcsDestinationUri,
        },
      };
      const features = [{type: 'DOCUMENT_TEXT_DETECTION'}];
      const request = {
        requests: [
          {
            inputConfig: inputConfig,
            features: features,
            outputConfig: outputConfig,
          },
        ],
      };
      console.log(processingOperation);
      var processingOperation;
      try {
        processingOperation = visionClient.asyncBatchAnnotateFiles(request); 
      } catch(processingError) {
        console.error(processingError);
        return processingError;
      }
      console.log(processingOperation);
```

The unhandled rejection is happening inside the Vision request try/catch block so there's nowhere further to debug for me, hope someone can help or is getting the same issue. This used to work without issue.

#### Environment details
- Google Cloud Functions, Node JS 10 (Beta) Platform
- NPM is building with  "@google-cloud/vision" : ">=1.2.0" so likely latest version

#### Steps to reproduce
- Using a GCF function on the Node JS 10 runtime, in a project with all api/auth enabled. 
- Make a request to visionClient.asyncBatchAnnotateFiles.

**Comment by timhj** [comment]
This issue is intermittent and because it's not catchable, it's resulting in 'out-of-whack' errors in the database. Without deploying anything differently, the code now 'just works' on the same retries.

**Comment by timhj** [comment]
The vision API request may have been failing behind the scenes and causing this issue from a request to output json into a non-existent GCS Bucket... As the issue is intermittent, I'm not sure. So will watch an see what happens.

**Comment by dusty** [comment]
I'm having the same problem with various libraries inside cloud functions.

This example is listening for a bucket onFinalize event and sending a single http task to a cloud tasks queue with the payload.

It seems that my function works fine when its first deployed. Then after some time, (perhaps) after it scales to zero and then is re-triggered it always fails until I redeploy it.

<img width="1175" alt="Screen Shot 2019-10-05 at 8 18 47 AM" src="https://user-images.githubusercontent.com/792/66257287-fde94f00-e74b-11e9-897c-cc2a7e0b64f8.png">

I'm using the nodejs10 runtime with `"@google-cloud/tasks": "^1.4.0"`.

Code
```
const { v2beta3 } = require('@google-cloud/tasks')
const client = new v2beta3.CloudTasksClient()

const queue = body => {
  return client.createTask({
    parent: process.env.QUEUE_URL,
    task: {
      httpRequest: {
        httpMethod: 'POST',
        url: process.env.TASK_URL,
        headers: { 'Content-Type': 'application/json' },
        body: Buffer.from(JSON.stringify(body)),
        oidcToken: { serviceAccountEmail: process.env.SERVICE_ACCOUNT_EMAIL }
      }
    }
  })
}

exports.default = async file => {
  await queue(file)
  console.info(`DONE: ${file.name} queued`)
}
```

**Comment by merlinnot** [comment]
I also experience this issue for quite some time now. It's happening in my project all the time, let me know if I can be of any help to debug it.

**Comment by bcoe** [comment]
@merlinnot what type of authentication are you using in your project, and what APIs specifically.

**Comment by merlinnot** [comment]
Firestore, BigQuery, Debugger, ...

Given the stack traces, the error observed originates here: https://github.com/googleapis/google-auth-library-nodejs/blob/944e2aa62a61c253ba153f49590d7416585c64eb/src/auth/googleauth.ts#L291-L296

As you can see, it is thrown if and only if the value of `isGCE` variable is falsy. The value is a result of a call of `_checkIsGCE ` function: https://github.com/googleapis/google-auth-library-nodejs/blob/944e2aa62a61c253ba153f49590d7416585c64eb/src/auth/googleauth.ts#L311-L316

This function in turn calls `isAvailable` function from `google-metadata` library:
https://github.com/googleapis/gcp-metadata/blob/25bc11657001cb6b3807543377d74bafe126ea62/src/index.ts#L121-L142

As you can see, it depends on `metadataAccessor ` function:
https://github.com/googleapis/gcp-metadata/blob/25bc11657001cb6b3807543377d74bafe126ea62/src/index.ts#L49

This function makes an HTTP request to http://169.254.169.254/computeMetadata/v1/ here:
https://github.com/googleapis/gcp-metadata/blob/25bc11657001cb6b3807543377d74bafe126ea62/src/index.ts#L66

I see no other way for this error to occur other than a requests to this service fail.

**Comment by merlinnot** [comment]
I'm currently redeploying all of the functions with additional logging enabled (`DEBUG_AUTH`). Will post here as soon as I have a hit.

**Comment by merlinnot** [comment]
In the last 24 hrs I had 71,092 occurrences of this error, but it was last seen 5 hrs ago... I thought I'll be able to provide you more information straight away, this error used to happen all the time.

**Comment by bcoe** [comment]
@merlinnot as you noticed, I've deployed a version of `gcp-metadata` with a debug option. I'd double check that your package-lock.json has `gcp-metadata@3.2.0`, at which point we should get a better picture of what's happening the next time you run into issues.

**Comment by davedc** [comment]
@bcoe Experiencing the same issue on a few functions on our side.

I can confirm @merlinnot's suspicion that requests to the metadata service is failing.

<img width="1829" alt="Screen Shot 2019-10-11 at 6 22 04 pm" src="https://user-images.githubusercontent.com/7055057/66632494-52ad2e00-ec54-11e9-86c9-f8bd30dbc2eb.png">

**Comment by merlinnot** [comment]
It's back :)

```
{ FetchError: network timeout at: http://metadata.google.internal./computeMetadata/v1/instance
at Timeout.<anonymous> (/srv/functions/node_modules/node-fetch/lib/index.js:1448:13)
at ontimeout (timers.js:436:11)
at tryOnTimeout (timers.js:300:5)
at listOnTimeout (timers.js:263:5)
at Timer.processTimers (timers.js:223:10)
message:
'network timeout at: http://metadata.google.internal./computeMetadata/v1/instance',
type: 'request-timeout',
config:
{ url:
'http://metadata.google.internal./computeMetadata/v1/instance',
headers: { 'Metadata-Flavor': 'Google' },
retryConfig:
{ noResponseRetries: 0,
currentRetryAttempt: 0,
retry: 3,
retryDelay: 100,
httpMethodsToRetry: [Array],
statusCodesToRetry: [Array] },
responseType: 'text',
timeout: 3000,
params: [Object: null prototype] {},
paramsSerializer: [Function: paramsSerializer],
validateStatus: [Function: validateStatus],
method: 'GET' } }
```

Here's a timeline for the last 30 days:
<img width="696" alt="Screen Shot 2019-10-11 at 11 38 13" src="https://user-images.githubusercontent.com/10390061/66641798-b7986200-ec1b-11e9-9736-85d48ec6c9d9.png">

And for the last 7 days:
<img width="697" alt="Screen Shot 2019-10-11 at 11 38 25" src="https://user-images.githubusercontent.com/10390061/66641811-bebf7000-ec1b-11e9-9043-7b5c5d784f5f.png">

**Comment by BluebambooSRL** [comment]
Same problem...

**Comment by bcoe** [comment]
@merlinnot @BluebambooSRL thank you, this gives us some valuable forensic information for the engineering team :+1:

**Comment by MorganGallant** [comment]
It seems that I am getting a similar issue on some of my cloud functions (Firebase) while attempting to access a Firestore database:
```
Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
    at GoogleAuth.getApplicationDefaultAsync (/srv/node_modules/google-auth-library/build/src/auth/googleauth.js:161:19)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:229:7)
```

And yes, I can confirm from my (limited) testing that a re-deploy does temporarily solve the problem. I appreciate @timhj for posting an issue about this, since it seems to be happening to a lot of people and I am hopeful this can be resolved quickly.

**Comment by smashah** [comment]
Any remedy for this? I'm experiencing it in same circumstance (vision API in GCF (Node v8))

**Comment by merlinnot** [comment]
I think it's not related to dependencies. In my case redeployments also change the behavior of these errors: sometimes I have more, sometimes I have less (see the chart above), where the number of executions per day is rather stable.

I wild guess would be that it just depends on which node in the underlying infrastructure the code lands? Maybe a re-re-deploying would help in your case?

**Comment by smashah** [comment]
@merlinnot Yes I just redeployed it with ^@google-cloud/vision@1.5.0 (before it was 1.4.0) and it started working again.

**Comment by davedc** [comment]
Sadly, at `europe-west2` seems like all underlying nodes have this issue? Redeploying the function a bunch of times has not really alleviated things for us.

**Comment by antonioallen** [comment]
Unfortunately, I'm running into this issue as well. It's happening pretty consistently for me at the moment. It just started after a recent full deploy of all my cloud functions. I'm receiving `Error: Could not load the default credentials.` followed by `Unhandled error Error: Can't set headers after they are sent.`(mostly likely from "Ignoring exception from a finished function" ). A bit out of ideas on this one. I'll keep poking at it.

For me it's happening at:
`GoogleAuth.getApplicationDefaultAsync (/srv/node_modules/@google-cloud/logging/node_modules/google-auth-library/build/src/auth/googleauth.js:161:19)`

<img width="1193" alt="Screen Shot 2019-10-19 at 9 28 47 AM" src="https://user-images.githubusercontent.com/11503003/67148332-77f0fc00-f253-11e9-8097-82bcd970ef9e.png">
<img width="1193" alt="Screen Shot 2019-10-19 at 9 26 47 AM" src="https://user-images.githubusercontent.com/11503003/67148333-77f0fc00-f253-11e9-97f5-de118ba7b1a4.png">

**Comment by antonioallen** [comment]
Commenting out all logging `logger.debug()` within the functions seems to fix the issue for me. But... no logs. Wonder why the auth is failing for it now.

**Comment by edi** [comment]
Same here. I have more apps .. none of them are failing with this error.

But one of them ( after recently upgrading my functions dependencies ), started resulting in the same error as above.

Weird thing is, given there are two functions, only one triggers this error.

I have no outside libraries or network requests, no APIs being used, simple firestore document triggers and updates. 

So while I’m using the latest version of everything, only one function out of the two is randomly failing. 

Had 11 fails during past 5 days. My client is losing revenue based on those fails though, so it’s a bit worrying.

**Comment by ollydixon** [comment]
Also seeing the same error because we are not running on a GCP environment but running the admin service. Not sure how to get this working. Was working yesterday..

We really don't want to migrate to cloud functions because of the extremelty slow boot times for every function.

Inside googleauth.js I can see the folllowing:

```
async _checkIsGCE() {
        if (this.checkIsGCE === undefined) {
            this.checkIsGCE = await gcpMetadata.isAvailable();
        }
        return this.checkIsGCE;
    }
```

Any idea if we could just bypass this?

**Comment by bcoe** [comment]
@ollydixon this thread is specifically related to authentication issues with cloud functions, which I think is potentially related to something specifically happening within this environment.

Could I bother you to open a new issue, with more specifics about the environment you're running in, and the steps you're using to bootstrap your application.

**Comment by bcoe** [comment]
@edi, @davedc, @smasha :wave: sorry about your frustration, I've raised an internal issue with the Cloud Functions folks (which is why this is labeled `external`), and am going to follow up again today.

**Comment by bcoe** [comment]
There's an internal issue with the GCF folks that has been updated throughout the day, we're trying to get the root cause of the timeouts that occur attempting to connect to the metadata server (this is what's in turn resulting in the credentials issue).

A potential workaround for folks, would be to create a service account, rather than relying on the default credentials:

https://cloud.google.com/docs/authentication/getting-started

:point_up: this requires that the credentials are available in a file on disk, so you would need to either use the API to deploy your project, or use the file upload option (rather than the inline editor).

At which point you would set your credentials to `GOOGLE_APPLICATION_CREDENTIALS=./my-service-account.son`.

I understand this workaround is suboptimal, and we are continuing to dig into things on our end.

**Comment by Dara-To** [comment]
I have the same problem on my new MacBook. I am able to run my functions locally on my old MacBook and iMac at the office. All using the same Firebase Functions project. So the default credential works on admin.initializeApp. I don't want to change the initialization code to use the workaround for 1 device. I will wait for the permanent fix. Please let us know when it will be fixed. Thanks.

**Comment by timhj** [comment]
Sounds like it might be a different issue to this one. I'd make sure you've
got the appropriate environment variable set with the location of your json
credentials first.

There's no implicit default credentials on a new Mac.

**Comment by bcoe** [comment]
@Dara-To I believe @timhj is correct, you will want to go through the steps outlined here:

https://cloud.google.com/docs/authentication/getting-started

To setup authentication on your new laptop; if you bump into issues, please feel free to open up an issue here.

**Comment by neelance** [comment]
I just saw the same error with one of our webhook functions after redeploying it. Special about this function was that it sent the response early with `resp.sendStatus(202);` and did the heavy lifting afterwards. This caused the function to already log `Function execution took X ms, finished with status code: 202` early. My guess is that the function then tried to fetch the default credentials, but couldn't get them because the execution was already marked as finished.

I am now writing the webhook's payload into the database before returning 202 and then using a separate cloud function to process it asynchronously. This made the error go away.

**Comment by Tebro** [comment]
This started happening to one of my functions now. Is there any news regarding a fix? We are using Cloud Build for function deployment and getting a json file embedded into the build without having it in git is non-trivial.

**Comment by bcoe** [comment]
@Tebro I see a few internal threads regarding timeout issues on the metadata server, there was one rollout today, but it looks like the issue is still periodically happening.

I will continue to keep this thread updated.

**Comment by aldobaie** [comment]
Hi guys, 
I ran into the same problem myself. Sometimes the function works and many times it would through the `Error: Could not load the default credentials` error.
The problem I believe have been solved by watching for the Callbacks. You have to keep the function running until the callbacks have been called using the `await` and `async` prefixes. 

Firebase Cloud Functions don't allow the access to the processor through callbacks once it's been terminated! That's why we get the `Error: Could not load the default credentials` error.

So, whenever you have a `.then() function` prefix it with `await` and prefix the function it's inside it with `async` and prefix any call to the function with `await`.
```
async function registerUser(..) {
    ...
    await admin.firestore().collection(..)...
    ...
}
```

I hope this helps you out!

**Comment by edi** [comment]
@aldobaie Not the case. 

It's just a request timeout, when it happens. Seems like the SDK can't each one of the credentials endpoints ( eg. 169.254.169.254 ).

Given the fact that it happens randomly ( for the same code snippets ) means it's not code-oriented.

The workaround (until they sort it) is to load your service account credentials manually like this:

```
admin.initializeApp({
	credential: admin.credential.cert(credentials),
	databaseURL: 'https://PROJECT_ID.firebaseio.com'
})
```

**Comment by Dara-To** [comment]
I think I got it. I installed the new version of the Firebase CLI firebase-tools on my new laptop whereas my iMac is still on an older version. I will confirm the version on my iMac when I get back to the office. I believe pass a certain version, they have removed the default credential so that `admin.initializeApp(functions.config().firebase)` wouldn't work anymore.

My next question is which service account should I generate a key for? I am so confused with the many options stated in the documentations. My backend project is only for cloud functions, no hosting. Any clarifications would be appreciated.

**1. Firebase service account**  

In the Firebase console, you could generate a key for the Firebase Admin SDK.
  
https://console.firebase.google.com/project/<project_id>/settings/serviceaccounts/adminsdk  

**2. App Engine default service account**

According to this doc,  it says to create a key for the App Engine default service account. 

https://firebase.google.com/docs/functions/local-emulator  https://console.cloud.google.com/iam-admin/serviceaccounts?project=<project_id>&authuser=0   

**3. New service account**  

According to this doc (provided by an earlier message in this thread), it says to set up a new service account and set the role as the project owner.  
https://cloud.google.com/docs/authentication/getting-started#auth-cloud-implicit-nodejs

**Comment by bcoe** [comment]
@timhj  @aldobaie @Tebro there have been some stability fixes deployed internally, are you continuing to see these issues?

**Comment by bcoe** [comment]
@Dara-To I'm glad you've made some progress :+1: could I bother you to open a new issue with your questions, this thread is related to specific issues were were seeing with cloud functions not loading default credentials (I don't want to lose your questions in the shuffle).

**Comment by Dara-To** [comment]
Sure thing, I can email Firebase support. Thanks

**Comment by bcoe** [comment]
@Dara-To happy to have a tracking issue here too; starting a conversation with Firebase is probably also worthwhile, given they'll have more specific expertise.

**Comment by timhj** [comment]
@bcoe - Still happens intermittently but not frequently - ![4 instances in the past 24hrs](https://user-images.githubusercontent.com/3896211/68632467-9a5ff780-0529-11ea-8e97-bae9f355dbde.png), the tight grouping of errors within microseconds of each other agree with it being a quick connection dropout issue. 

How recently a redeploy does the fix need to have been included?

**Comment by NawarA** [comment]
Definitely experiencing this, particularly when I have about >1,400 instances active, seems some sputter and die with this error

**Comment by bcoe** [comment]
@timhj the deploy I believe was in the past week, mind sharing with me the project identifier (feel free to send it by email `bencoe [at] google.com`.

It sounds promising that this error has become a rare occurrence, makes me think the upstream issue has at least been partially addressed.

**Comment by Tebro** [comment]
Have not seen it in a while, but the project does not have high activity at this point.

**Comment by seriousManual** [comment]
Just to add on this, we're seeing this error as well in the context of `logging-winston`  (https://www.npmjs.com/package/@google-cloud/logging-winston).

This happens almost everytime the cloud function is redeployed (e.g. after an update). The function is running but logging does not work.
I'm writing *almost* here, because from time to time it is actually working, but not consistently.

**Comment by smashah** [comment]
Just for context, I deployed my functions - it did not work. I deployed them again and it's been working ever since. Very odd.

**Comment by zoharsanchez** [comment]
Redeploying works, but it's trail and error. We get the error, redeploy and check that. We have a credential file, but it is encrypted in Storage and we decrypt it when the function is run. Really odd behavior indeed.

**Comment by seriousManual** [comment]
> Note, the workaround of specify an explicit service account, using GOOGLE_APPLICATION_CREDENTIALS should do the trick for the time being as a workaround.
Hi @bcoe 
would that be an env variable at deploytime or a file in the root of the function?
Thanks!

**Comment by arfnj** [comment]
> > Note, the workaround of specify an explicit service account, using GOOGLE_APPLICATION_CREDENTIALS should do the trick for the time being as a workaround.
> 
> Hi @bcoe
> would that be an env variable at deploytime or a file in the root of the function?
> Thanks!

@seriousManual I used this workaround and haven't had any errors since.  I uploaded the secret in the root of the function and then declared the variable at the start of my code:

`process.env.GOOGLE_APPLICATION_CREDENTIALS = './my-secret.json;`

**Comment by NawarA** [comment]
Good workaround! 

Though the good folks on the team need to dig into the root cause since their software is inconsistent at scale, which means there's a deeper root cause to discover and patch :)

**Comment by runelk** [comment]
We had the same issue in our Cloud Functions in a Firebase project, but the workaround proposed by  @arfnj seems to fix it. Are there any potential security issues to take note of when doing it this way?

**Comment by arfnj** [comment]
> We had the same issue in our Cloud Functions in a Firebase project, but the workaround proposed by @arfnj seems to fix it. Are there any potential security issues to take note of when doing it this way?

I appreciate the kudos, @NawarA and @runelk, but that's not my workaround!  The esteemed @bcoe suggested it on October 21st and I was just sharing my implementation of it to help answer @seriousManual's question.  🙂

**Comment by latifs** [comment]
@arfnj,
not sure if you can answer this but am thinking about implementing this option too.

How would you dynamically switch the `secret.json` file if you had a prod and dev environment?
Thanks

**Comment by Dara-To** [comment]
I tried to use the workaround on my new MPB laptop with setting `export GOOGLE_APPLICATION_CREDENTIALS="path/to/key.json"` on my shell session as suggested by @bcoe but it didn't work. I even updated all the firebase dependencies. Why does this have to be so complicated?

Is it because of this message on my terminal?

`The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.`

**Comment by bcoe** [comment]
@Dara-To another option for you would be installing the [gcloud command line tool](https://cloud.google.com/sdk/gcloud/), running `gcloud auth login` (to login to your account), and then also making sure you set a default project ID, with `gcloud config set project`.

You can also run your application like so:

```
GOOGLE_APPLICATION_CREDENTIALS="path/to/key.json" node my-app.js
```

If you continue to have issues please feel free to open a separate issue in this library, or to reach out to support for the specific product you're using 👍 

Your issue is I believe unrelated to this thread however, which relates to folks running applications in a GCP environment (not on their local machine).

**Comment by GuanacoDevs** [comment]
I have been having this issue for awhile now. with `"googleapis": "^45.0.0"` in firebase cloud functions.
I do use it to verify purchases/subscriptions. I do have noticed that the functions crashes when gets triggered, after a long time of not being triggered. After the first crash, don't seem to happen again, with tests purchases...
I get my credentials like this:
`const authClient = new google.auth.JWT({
    email: functions.config().verifier.email,
    key: functions.config().verifier.key,
    scopes: [functions.config().verifier.scope]
});`
I used to have it hardcoded, then found out about environment variables and started using them.
Hte `authClient` is inside the pubsub triggered function, as is the only function where I use it, I'm taking it out of the function, maybe is related to a cold start!?
For me is a big deal because when this function crashes, I can't deliver the goods and I don't think that's a good thing for my users.

**Comment by NawarA** [comment]
I'm using the default credentials built into the cloud functions environment. That's supposed to, and has historically, worked right out of the box. Its supposed to be supported, so that means it should work.

Here's the log of what one of these errors looks like:

```
{
 insertId: "000000-b484b5cc-71b5-4d9e-987d-2420d38bfc5c"  
 labels: {…}  
 logName: "projects/bytez-web-services/logs/cloudfunctions.googleapis.com%2Fcloud-functions"  
 receiveTimestamp: "2019-12-01T06:22:38.408232522Z"  
 resource: {…}  
 textPayload: "Exception from a finished function: Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information."  
 timestamp: "2019-12-01T06:22:36.493Z"  
 trace: "projects/bytez-web-services/traces/f9014e52bc5aa4176b1b602ea214b654"  
}
```

race condition?

**Comment by stari4ek** [comment]
I'm getting this one when processing lots of `firestore` documents with 100 promises in parallel per-batch. forcing initialization from service account JSON instead of default one - fixes the issue

**Comment by NawarA** [comment]
Yeah, I did the same.

I abandoned using the default creds that come with cloud functions since they're the root cause of the issue (being inconsistently available). Uploading a service account and directly pointing to it works.

I still think the awesome G-team should find the root cause and release a patch

**Comment by moblizeit** [comment]
Same issue our production instance hence the customers are affected.

**Comment by bcoe** [comment]
@moblizeit @stari4ek are you both using Cloud Functions? I have a recommendation I am going to make an effort to deploy this week.

The workaround of referencing the service account directly should address the issue, but we are making efforts to make address the underlying issues.

**Comment by stari4ek** [comment]
Yes, I get it only when do highly concurrent firestore processing from firebase functions (reading/writing non-overlapping documents from different collections)

**Comment by moblizeit** [comment]
@bcoe  we are using Firebase functions, firebase DB  and google ml and vision apis overall. 

The error came from firebase function while trying to run OCR apis

**Comment by kosukekurimoto** [comment]
The problem continues.
<img width="1283" alt="スクリーンショット 2019-12-21 2 27 27" src="https://user-images.githubusercontent.com/27758476/71277935-8c5b8d00-2399-11ea-9c73-494680db1e19.png">

**Comment by bcoe** [comment]
@kosukekurimoto a few questions for you:

1. have you redeployed your functions recently (I've gotten reports that `gcp-metadata@3.3.0`, which was recently released, makes these errors go away in the reproduction we'd created).
2. is this happening all of the time, or is it an intermittent error?

**Comment by kosukekurimoto** [comment]
> 1. 
The function was deployed today.
> 2.
It always happens. This can be avoided by setting an environment variable.

**Comment by aragalie** [comment]
Came here to leave a suggestion: you might have promise callbacks in your code not managed properly (using await). I faced the same auth error as above, but then found this answer https://stackoverflow.com/a/58779000/8500237 ; after rechecking my code, I found that awaiting the promises fixed also the functions errors :)  Hope it helps someone else.

**Comment by danielmbarlow** [comment]
Thanks @aragalie I suspect you are right. I have the same in my logs when using Google Cloud PubSub in a Google Cloud function:
`Exception from a finished function: Error: Could not load the default credentials`

In my case the Google Cloud Function had an `await`, but I had Promises in the the stack that weren't being returned to the Google Cloud Function entrypoint.

**Comment by aragalie** [comment]
> Thanks @aragalie I suspect you are right. I have the same in my logs when using Google Cloud PubSub in a Google Cloud function:
> `Exception from a finished function: Error: Could not load the default credentials`
> 
> In my case the Google Cloud Function had an `await`, but I had Promises in the the stack that weren't being returned to the Google Cloud Function entrypoint.

happy it helped! @bcoe perhaps the error log message could be improved for this case?

**Comment by NayanJD** [comment]
> Came here to leave a suggestion: you might have promise callbacks in your code not managed properly (using await). I faced the same auth error as above, but then found this answer https://stackoverflow.com/a/58779000/8500237 ; after rechecking my code, I found that awaiting the promises fixed also the functions errors :) Hope it helps someone else.

Helped me too. Thanks a lot!!!

**Comment by bcoe** [comment]
@timhj I believe we've addressed some load issues that were likely a major cause of these intermittent connectivity issues, how have your logs looked? I'm wondering if we can close this out.

**Comment by timhj** [comment]
Hi Ben,

There's nothing in the logs since 23/12 when there were 15 or so
occurrences about 21:19pm. That being said, Christmas break has seen pretty
low usage.

When did the load changes get rolled out?

Cheers,
Tim.

PS. Happy New Year.

**Comment by moblizeit** [comment]
I saw it just last night see the screenshot https://www.dropbox.com/s/c3yold4eyrkvtc1/Screenshot%202020-01-02%2017.44.13.png?dl=0

i am 100% sure it's not fixed yet. just was on a christmas break

**Comment by timhj** [comment]
@bcoe - FWIW similar to @moblizeit's example, this issue happens on interactions with the Cloud Vision API:
![Screen Shot 2020-01-03 at 12 51 28 PM](https://user-images.githubusercontent.com/3896211/71707848-e6436480-2e27-11ea-90c5-8a0b7c2163a9.png)
```
textPayload: "Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
    at GoogleAuth.getApplicationDefaultAsync (/srv/functions/node_modules/@google-cloud/vision/node_modules/google-auth-library/build/src/auth/googleauth.js:161:19)
```

Have the team been looking into an issue specific to the Vision API from cloud functions or all APIs in general?

**Comment by bcoe** [comment]
@timhj I doubt that it relates to specific APIs, and am pretty convinced that it is caused by intermittent issues accessing the metadata server that we use to detect environment in google cloud functions. There's another PR I'll merge Monday, which adds additional caching:

https://github.com/googleapis/google-auth-library-nodejs/pull/858

My hope is that this will further improve the situation.

**Comment by idudinov** [comment]
I was observing the issue on the last week of 2019, a function that worked OK a while before began failing with the default credentials error on a Firestore request, but no changes has been made. So I just re-deployed the function with the same source code and it just helped.

**Comment by kkogovsek** [comment]
It's also happening to us using firebase functions. Most of the time it's random. At some times every function execution fails, but most of the time it works. Is there a way to try/catch it or are the references lost?

**Comment by DallasHoff** [comment]
Same here. The error is occurring in _recently updated_ Firebase Cloud Functions that access Cloud Firestore or the Firebase Realtime Database using the admin SDK.

**Comment by cmditch** [comment]
Also experiencing this recently in our Firebase Cloud Functions.

**Comment by psyh** [comment]
It is so stupid!!
Also using Firebase Cloud Function - didn't change anything and it started to happen several minuted then I redeployed the project without any changes and it despaired. Now it started to happen again! I tried all the recommendation from this page and from other. But it is just so stupid!

**Comment by vmal** [comment]
Same problem! I have two environments where I am using google vision and one works fine but the other is always giving me the same error again and again.

**Comment by martinatcode** [comment]
@bcoe This issue remains unsolved and I face the error daily. Do you have any update(s) regarding the progress on this issue?

**Comment by bcoe** [comment]
@martinatcode are you using firebase functions as well?

**Comment by martinatcode** [comment]
@bcoe Yes, I'm using Firebase Functions as well.

**Comment by adamworrall** [comment]
I didn't see this error for approximately a month or so but it's returned. I don't see how unresolved promises would cause the issue, as mentioned by a few people above. I haven't seen any documentation anywhere that warns about unresolved promises being a problem other than not guaranteeing execution after the main function has returned. The bigger concern for me is that it seems to kill the whole function until the next deployment when this problem shows up.

**Comment by jamesearl** [comment]
It's returned for me as well. I've tried using the default application credentials, and I've tried using my own service account, as mentioned earlier in the thread.

With default creds, I still get `Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information`.

With my own provided credentials, I get `Error: request to https://www.googleapis.com/oauth2/v4/token failed, reason: Client network socket disconnected before secure TLS connection was established`.

I can successfully run my function about 20% of the time.

**Comment by BrodaNoel** [comment]
Yeap. It's back.
I also have random non-working-executions (80% of the time)

**Comment by rthotakura-wp** [comment]
@bcoe 
It happened for one of our functions that is configured to use credentials explicitly by setting `GOOGLE_APPLICATION_CREDENTIALS` environment variable.
I also saw `Error: function crashed out of request scope Function cannot be executed` errors.

<img width="1348" alt="default_errors" src="https://user-images.githubusercontent.com/55157573/72830138-eff50680-3c34-11ea-91d1-8cc35d5339de.png">

**Comment by 27Bslash6** [comment]
Data point: Seeing the same issue when adding the logging-winston library in GCF - https://www.npmjs.com/package/@google-cloud/logging-winston .  Without this the winston wrapper, the functions execute just fine.

We are *not* using Firebase functions, but we are using functions-framwork and trace-agent:

```
    "@google-cloud/functions-framework": "^1.3.2",
    "@google-cloud/logging-winston": "^3.0.5",
    "@google-cloud/pubsub": "^1.3.0",
    "@google-cloud/storage": "^4.2.0",
    "@google-cloud/trace-agent": "^4.2.5"
```

Is this due to the library sending logs to Stackdriver asynchronously, and the function is exiting before the promise is resolved?

![image](https://user-images.githubusercontent.com/2221076/72853477-93155100-3d05-11ea-8297-b2cf6394dc4f.png)

**Comment by jamesearl** [comment]
To be clear: explicitly providing service account credentials as mentioned both here and in googleapis/nodejs-logging-winston#443 is not functioning as a workaround for this issue.

**Comment by jketcham** [comment]
Just to chime in here, I started experiencing this with almost every invocation of a particular function in development in the past 24 hours (and have experienced it in the past as well, just not recently).

**Error**:
```
{
 insertId: "000000-fe445c77-1d32-4e80-b3b7-79ddefbf3769"  
 labels: {
  execution_id: "l70qcfotjepv"   
 }
 logName: "projects/paths-development-260120/logs/cloudfunctions.googleapis.com%2Fcloud-functions"  
 receiveTimestamp: "2020-01-22T20:09:20.477496971Z"  
 resource: {
  labels: {
   function_name: "locationDocuments"    
   project_id: "paths-development-260120"    
   region: "us-central1"    
  }
  type: "cloud_function"   
 }
 severity: "ERROR"  
 textPayload: "Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
    at GoogleAuth.getApplicationDefaultAsync (/srv/functions/node_modules/google-auth-library/build/src/auth/googleauth.js:159:19)
    at process._tickCallback (internal/process/next_tick.js:68:7)"  
 timestamp: "2020-01-22T20:09:18.906Z"  
 trace: "projects/paths-development-260120/traces/ed6bf6b2bd4973dea8445279c18bc666"  
}
```

**Context**:
This is a function that queries BigQuery and inserts ~30k documents into Firestore. They are batched in groups of 250/500 writes at a time. Timeout is 300 seconds on the function.

**Deps**:
```
"@google-cloud/bigquery": "^4.4.0",
"@google-cloud/firestore": "^2.6.1",
"@google-cloud/functions-framework": "^1.3.2",
```

**Comment by jketcham** [comment]
Just to chime in here, I started experiencing this with almost every invocation of a particular function in development in the past 24 hours (and have experienced it in the past as well, just not recently).

**Error**:
```
{
 insertId: "000000-fe445c77-1d32-4e80-b3b7-79ddefbf3769"  
 labels: {
  execution_id: "l70qcfotjepv"   
 }
 logName: "xxx/logs/cloudfunctions.googleapis.com%2Fcloud-functions"  
 receiveTimestamp: "2020-01-22T20:09:20.477496971Z"  
 resource: {
  labels: {
   function_name: "xxx"    
   project_id: "xxx"    
   region: "us-central1"    
  }
  type: "cloud_function"   
 }
 severity: "ERROR"  
 textPayload: "Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
    at GoogleAuth.getApplicationDefaultAsync (/srv/functions/node_modules/google-auth-library/build/src/auth/googleauth.js:159:19)
    at process._tickCallback (internal/process/next_tick.js:68:7)"  
 timestamp: "2020-01-22T20:09:18.906Z"  
 trace: "xxx"  
}
```

**Context**:
This is a function that queries BigQuery and inserts ~30k documents into Firestore. They are batched in groups of 250/500 writes at a time. Timeout is 300 seconds on the function.

**Deps**:
```
"@google-cloud/bigquery": "^4.4.0",
"@google-cloud/firestore": "^2.6.1",
"@google-cloud/functions-framework": "^1.3.2",
```

**Comment by jamesbenns** [comment]
I'm also using Firebase Cloud functions and experiencing this issue only for functions connecting to the Vision API. I have to just keep redeploying until it miraculously works (about 1 in 10 times)! 🤬

**Comment by bcoe** [comment]
@jamesbenns @jketcham @adamworrall, and others, could I bother you to try setting the environment variable `DETECT_GCP_RETRIES` to `3`? This will make the library retry when detecting whether the environment is GCP.

Along with this, could you also set the environment variable `DEBUG_AUTH` to `true`, this will provide more detailed information about the error that occurred detecting the environment (I would love to see this detailed information).

**Comment by jamesearl** [comment]
@bcoe 

Those env vars appear to have no effect when specifying my own service account.

Using default credentials:
```
{ FetchError: network timeout at: http://169.254.169.254/computeMetadata/v1/instance  
    at Timer.processTimers (timers.js:223:10)  
    at Timeout.<anonymous> (/srv/functions/node_modules/node-fetch/lib/index.js:1448:13)  
    at ontimeout (timers.js:436:11)  
  message:  
   'network timeout at: http://169.254.169.254/computeMetadata/v1/instance',  
    at tryOnTimeout (timers.js:300:5)  
  type: 'request-timeout',  
    at listOnTimeout (timers.js:263:5)  
  config:  
   { url: 'http://169.254.169.254/computeMetadata/v1/instance',  
     headers: { 'Metadata-Flavor': 'Google' },  
     retryConfig:  
      { noResponseRetries: 0,  
        currentRetryAttempt: 0,  
        retry: 3,  
        retryDelay: 100,  
        httpMethodsToRetry: [Array],  
        statusCodesToRetry: [Array] },  
     responseType: 'text',  
     timeout: 3000,  
     params: [Object: null prototype] {},  
     paramsSerializer: [Function: paramsSerializer],  
     validateStatus: [Function: validateStatus],  
     method: 'GET' } }  

Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
    at Promise.all.catch (/srv/functions/node_modules/scramjet-core/lib/util/mk-transform.js:59:44)
    at process._tickCallback (internal/process/next_tick.js:68:7)
   caused by: 
     at GoogleAuth.getApplicationDefaultAsync (/srv/functions/node_modules/google-auth-library/build/src/auth/googleauth.js:159:19)
     at process._tickCallback (internal/process/next_tick.js:68:7)
   --- raised in DataStream(8) constructed ---
     at new PromiseTransformStream (/srv/functions/node_modules/scramjet-core/lib/util/promise-transform-stream.js:65:27)
     at new DataStream (/srv/functions/node_modules/scramjet-core/lib/data-stream.js:43:9)
     at DataStream.map (/srv/functions/node_modules/scramjet-core/lib/data-stream.js:186:26)
     at process$ (/srv/functions/process.js:59:35)
     at tryCatch (/srv/functions/node_modules/regenerator-runtime/runtime.js:45:40)
     at Generator.invoke [as _invoke] (/srv/functions/node_modules/regenerator-runtime/runtime.js:271:22)
     at Generator.prototype.(anonymous function) [as next] (/srv/functions/node_modules/regenerator-runtime/runtime.js:97:21)
     at tryCatch (/srv/functions/node_modules/regenerator-runtime/runtime.js:45:40)
     at invoke (/srv/functions/node_modules/regenerator-runtime/runtime.js:135:20)
     at /srv/functions/node_modules/regenerator-runtime/runtime.js:170:11
 Function execution took 17000 ms, finished with status: 'error'
```

Using a service account:
```
Error: request to https://www.googleapis.com/oauth2/v4/token failed, reason: Client network socket disconnected before secure TLS connection was established
     at Promise.all.catch (/srv/functions/node_modules/scramjet-core/lib/util/mk-transform.js:59:44)
     at process._tickCallback (internal/process/next_tick.js:68:7) 
   caused by: 
     at ClientRequest.<anonymous> (/srv/functions/node_modules/node-fetch/lib/index.js:1455:11) 
     at ClientRequest.emit (events.js:198:13) 
     at ClientRequest.EventEmitter.emit (domain.js:448:20) 
     at TLSSocket.socketErrorListener (_http_client.js:392:9) 
     at TLSSocket.emit (events.js:198:13) 
     at TLSSocket.EventEmitter.emit (domain.js:448:20) 
     at emitErrorNT (internal/streams/destroy.js:91:8) 
     at emitErrorAndCloseNT (internal/streams/destroy.js:59:3) 
     at process._tickCallback (internal/process/next_tick.js:63:19) 
   --- raised in DataStream(8) constructed --- 
     at new PromiseTransformStream (/srv/functions/node_modules/scramjet-core/lib/util/promise-transform-stream.js:65:27) 
     at new DataStream (/srv/functions/node_modules/scramjet-core/lib/data-stream.js:43:9) 
     at DataStream.map (/srv/functions/node_modules/scramjet-core/lib/data-stream.js:186:26) 
     at process$ (/srv/functions/process.js:59:35) 
     at tryCatch (/srv/functions/node_modules/regenerator-runtime/runtime.js:45:40) 
     at Generator.invoke [as _invoke] (/srv/functions/node_modules/regenerator-runtime/runtime.js:271:22) 
     at Generator.prototype.(anonymous function) [as next] (/srv/functions/node_modules/regenerator-runtime/runtime.js:97:21) 
     at tryCatch (/srv/functions/node_modules/regenerator-runtime/runtime.js:45:40) 
     at invoke (/srv/functions/node_modules/regenerator-runtime/runtime.js:135:20) 
     at /srv/functions/node_modules/regenerator-runtime/runtime.js:170:11 
 Function execution took 74607 ms, finished with status: 'error' 
```

**Comment by bcoe** [comment]
@jamesearl I do not believe that second error is coming from issues with the metadata server, could I bother you to open a new issue with this information? I think it's going to me something else happening in your environment, potentially networking rules that are blocking egres.

**Comment by jamesearl** [comment]
@bcoe I'm happy to open another ticket, but I strongly disagree that my comment is off-topic. 

You've suggested in multiple issues, including this one, that the workaround for the issue as originally stated is to provide service account credentials to the code running in a cloud function using the `GOOGLE_APPLICATION_CREDENTIALS` env var. 

The second error is what happens when I implement the workaround. How is that off-topic?

**Comment by bcoe** [comment]
@jamesearl this thread is attempting to track issues related to folks connecting to the metadata server, setting `GOOGLE_APPLICATION_CREDENTIALS` will work in GCP environments as a workaround. You are bumping into an issue in your environment, I think probably related to an unrelated edge-case. I would like to keep this thread focused on difficulties loading credentials from the metadata server, so that I can use it to concisely demonstrate the problems we're bumping into to folks internally.

**Comment by Cyganieszka** [comment]
I'm encountering same error. In my case I do Promise.all(list_of_promises), where each promise does http request(by request-promise). It works as expected if list_of_promises length is ~10, it fails with topic's error if array is bigger.
I double checked if I handle all promises correctly, but I'm not a JS programmer, so I keep in mind that could be my fault...

**Comment by glaforge** [comment]
I'm also experiencing this error.

In my context: I'm using Google Cloud Functions, with the Node.js 8 runtime, when calling the Cloud Vision API (with the latest 1.8.0 version of the @google-cloud/vision module).

```
Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information. at GoogleAuth.getApplicationDefaultAsync (/srv/node_modules/google-gax/node_modules/google-auth-library/build/src/auth/googleauth.js:160:19) at <anonymous> at process._tickDomainCallback (internal/process/next_tick.js:229:7)
```

My code is simply calling the Vision API:

```javascript
const request = {
        image: {source: {imageUri: `gs://${filebucket}/${filename}`}},
        features: [
            {type: 'LABEL_DETECTION'},
            {type: 'IMAGE_PROPERTIES'},
            {type: 'SAFE_SEARCH_DETECTION'}
        ]
    };

    const [response] = await client.annotateImage(request);
```

Although the trace doesn't show my own code, it's when I do the Vision API call that this error takes place.

I'm analysing some pictures, to get some information about that picture, and it doesn't seem to be triggered consistently. Some pictures work and go through, while others are failing with that error. Some pictures I tried before and worked, are failing now, while other brand-new pictures don't exhibit the problem.

**Comment by bcoe** [comment]
@glaforge @Cyganieszka @jamesbenns @BrodaNoel @psyh, I apologize for how frustrating this intermittent outage has been (_and for hiding your issue @jamesearl_).

I'm in the process of writing up a document, to communicate the severity of this problem to peers at Google. I have a favor to ask, could I ask folks to run their service with the environment variable:

```
DEBUG_AUTH=true
```

And to send me the error that you see in logs, either by responding in thread or by emailing (`bencoe [at] google.com`). This will hopefully give me the information I need to get more eyeballs on the problem.

**Comment by glaforge** [comment]
I've redeployed my (failing) function with the environment variable, but this morning, the problem has disappeared again. The fact it's intermittent is painful. But I'll keep monitoring if that error comes up again.

**Comment by glaforge** [comment]
Also where would I see the debug logs? What resources are they appearing under in stackdriver?
(I didn't see auth related logs in stackdriver)

**Comment by glaforge** [comment]
Actually, I've just seen a new occurrence of the problem.

At 14:12:30.749, I got the error reported above:
```Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information. at GoogleAuth.getApplicationDefaultAsync (/srv/node_modules/google-gax/node_modules/google-auth-library/build/src/auth/googleauth.js:160:19) at <anonymous> at process._tickDomainCallback (internal/process/next_tick.js:229:7)```

And before that, in the traces, I saw:

At 14:11:11.936, the trace:
```
{ FetchError: network timeout at: http://169.254.169.254/computeMetadata/v1/instance
    at Timeout.<anonymous> (/srv/node_modules/node-fetch/lib/index.js:1448:13)
    at ontimeout (timers.js:498:11)
    at tryOnTimeout (timers.js:323:5)
    at Timer.listOnTimeout (timers.js:290:5)
  message: 'network timeout at: http://169.254.169.254/computeMetadata/v1/instance',
  type: 'request-timeout',
  config: 
   { url: 'http://169.254.169.254/computeMetadata/v1/instance',
     headers: { 'Metadata-Flavor': 'Google' },
     retryConfig: 
      { noResponseRetries: 0,
        currentRetryAttempt: 0,
        retry: 3,
        httpMethodsToRetry: [Array],
        statusCodesToRetry: [Array] },
     responseType: 'text',
     timeout: 3000,
     params: {},
     paramsSerializer: [Function: paramsSerializer],
     validateStatus: [Function: validateStatus],
     method: 'GET' } }
```

And 400ms later at 14:11:12.336, I got:
```
{ FetchError: network timeout at: http://169.254.169.254/computeMetadata/v1/instance
    at Timeout.<anonymous> (/srv/node_modules/node-fetch/lib/index.js:1448:13)
    at ontimeout (timers.js:498:11)
    at tryOnTimeout (timers.js:323:5)
    at Timer.listOnTimeout (timers.js:290:5)
  message: 'network timeout at: http://169.254.169.254/computeMetadata/v1/instance',
  type: 'request-timeout',
  config: 
   { url: 'http://169.254.169.254/computeMetadata/v1/instance',
     headers: { 'Metadata-Flavor': 'Google' },
     retryConfig: 
      { noResponseRetries: 0,
        currentRetryAttempt: 0,
        retry: 3,
        httpMethodsToRetry: [Array],
        statusCodesToRetry: [Array] },
     responseType: 'text',
     timeout: 3000,
     params: {},
     paramsSerializer: [Function: paramsSerializer],
     validateStatus: [Function: validateStatus],
     method: 'GET' } }
```

Both seemed to use the IP address of the metadata server, and not the DNS name.

**Comment by adamworrall** [comment]
Metadata server having 169.254.x IP address in those traces is interesting?

**Comment by kmilo93sd** [comment]
any solution for this issue? i have problem with cloud functions trying to use firestore.

**Comment by glaforge** [comment]
@kmilo93sd do you get this problem all the time, consistently, or just once in a while, sporadically?
We're still missing a reproducible example, that fails all the time. Otherwise it's impossible to debug :-(

**Comment by kmilo93sd** [comment]
@glaforge well, in my last project i use firestore with the client sdk and i dont have problem. Now in my current project i get this issue all time with the admin sdk. 
I am using the production mode in firestore, so I think that to connect from my local environment to firestore I need to use a service account. By deploying these functions it get the credentials without problems.
I will have to use a service account and that's it, I don't know if the expected behavior is that, but it would be nice to use the credentials that I already have configured. Could you confirm if this behavior is expected when working with the mode of production from my local environment? Even if I'm already logged in with my firebase credentials.

**Comment by ilanbm** [comment]
Same here, trying to delete firestore collections and getting that error.
it works when launching on local serve, fails on firebase functions.

**Comment by SaschaHeyer** [comment]
I can reproduce the same behavior with BigQuery (since today). Cloud Function did not changed since 17th of January. 

Any solution?

**Comment by claytongulick** [comment]
Same issue. I'm ditching all stackdriver debugging and logging and just using system logs and slack integration because of this.

**Comment by mohshraim** [comment]
using cloud functions from a year, its my first time facing this problem..
what i did is just splitting some functionality to another  **help.js**  file other than **index.js** 
if help.js  have any work related to **DB operations (get,update...etc)** i got the mentioned error...   
so i keep the logic on help.js and move the update function to main.js and error **gone**!!!! 
Hope this help others 
still this a problem i dont have the correct answer for

**Comment by jhnnyk** [comment]
I was getting this error until I found this post: 
https://groups.google.com/forum/#!topic/firebase-talk/W4nB1Ykw7tM

returning the main chain seems to have resolved this error for me

**Comment by discobeta** [comment]
Try `export GOOGLE_APPLICATION_CREDENTIALS=~/.config/path-to-credential.json`

**Comment by bdaz** [comment]
+1, I've started seeing this in my Cloud Functions recently as well. It's sporadic.

**Comment by elihorne** [comment]
I've started to see this as well. The error began 10 days ago. 

Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
    at GoogleAuth.getApplicationDefaultAsync (/srv/node_modules/google-auth-library/build/src/auth/googleauth.js:160:19)
    at <anonymous>
    at process._tickDomainCallback (internal/process/next_tick.js:229:7)

**Comment by ivoberger** [comment]
We had this issue recently in 2 Firebase projects when trying to write to Firestore from a FB cloud function on nodejs 8. Also appeared without an obvious reason. In our case the reason seems to have been a cron job CF we added. After removing that the error didn't appear again.

**Comment by elihorne** [comment]
I have a job that has been behind a cron job for more than a year without issue. It requests a remote resource, writes to cloud storage, and then updates a database. This is where the error is occurring. It also uses firebase queue if that helps at all

**Comment by bcoe** [comment]
@elihorne could you share the basic structure of the code, I believe there's a chance we've debugged the root cause. If you're instantiating a client, e.g.,


```js
const {VisionClient} = require('@google-cloud/vision');
const vision = new Vision();
exports.myHandler = async (req, res) => {
  await vision.doSomething();
  res.send(result);
}
```

instead do this:

```js
const {VisionClient} = require('@google-cloud/vision');
let vision;
exports.myHandler = async (req, res) => {
  if (!vision) {
    vision = new Vision();
  }
  await vision.doSomething();
  res.send(result);
}
```

☝️ when you `new` client libraries, this causes an HTTP request to be made to detect the environment, on cold starts this request might not be made with CPU/memory available.

**Comment by elihorne** [comment]
@bcoe is the important part:

``if (!vision) {
    vision = new Vision();
  }
``

or that it's wrapped in an async function?

**Comment by bcoe** [comment]
@elihorne it's important that the async work has finished before you call `res.send`. Both parts are important:

1. it's important no async work happens in the global scope (`new Vision()` currently creates async work, we're fixing this).
2. as soon as `res.send()` is called, you may lose CPU/memory, so hold off on that until critical async work is finished.

**Comment by bcoe** [comment]
@elihorne please let me know if this works, because it's good confirmation that we're on the right track for a fix.

**Comment by DallasHoff** [comment]
@bcoe I was calling `new CloudTasksClient()` in the global scope. I have just made the suggested change, and I will return with an update on whether the errors stop happening.

**Comment by juloko** [comment]
Same here, with Firestore:
`>  Got an error Error: Could not load the default credentials. Browse to https://cloud.google.com/docs/authentication/getting-started for more information.
>      at GoogleAuth.getApplicationDefaultAsync (C:\Users\Ocara\Desktop\Traderox\functions\node_modules\google-auth-library\build\src\auth\googleauth.js:161:19)
>      at processTicksAndRejections (internal/process/task_queues.js:97:5)
>      at async GoogleAuth.getClient (C:\Users\Ocara\Desktop\Traderox\functions\node_modules\google-auth-library\build\src\auth\googleauth.js:503:17)
>      at async GrpcClient._getCredentials (C:\Users\Ocara\Desktop\Traderox\functions\node_modules\google-gax\build\src\grpc.js:108:24)
>      at async GrpcClient.createStub (C:\Users\Ocara\Desktop\Traderox\functions\node_modules\google-gax\build\src\grpc.js:229:23)
127.0.0.1 - - [21/Feb/2020:14:14:50 +0000] "GET /getMarket HTTP/1.1" 504 - "-" "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.116 Safari/537.36"`

**Comment by bcoe** [comment]
@juloko this looks like its a program you are running on your local machine, and is not running in Cloud Functions? (I'm gathering this based on the `C:\Users\Ocara`.

Have you followed the [instructions here](https://cloud.google.com/docs/authentication/getting-started) for authentication?

**Comment by DallasHoff** [comment]
@bcoe Update: Since making the change discussed above, I have not experienced the error. 🙌

**Comment by calclavia** [comment]
I'm running into this issue when using Firebase cloud functions. It seems to happen on some cloud functions that are deployed, but goes away if I re-deploy the functions.

**Comment by omersadika** [comment]
I'm getting the same issue for firebase functions. It's randomly happening sometimes after I deploy. The way to solve it is by redeploying the function / deleting and deploying until it's working. This is destroying productivity for my team, we're spending hours on dealing with it.

**Comment by kthaas** [comment]
We are also having this issue. We have about 65 Firebase / Google cloud functions in deployment. Our services will randomly fail at some points with this failure to load default credentials issue and the only remediation we have found is a redeployment.

**Comment by vmal** [comment]
This issue has been there for a long time and its affecting our production system so much. Important demos are being missed because this issue randomly shows up. Today has been way worse where even after multiple deletion and deploys i got the same error. This is so ******* frustrating!

**Comment by bcoe** [comment]
@vmal @kthaas @omersadika @calclavia, we believe we've isolated this problem to creating a client in the global scope outside of your function [see: function execution lifetime](https://cloud.google.com/functions/docs/concepts/exec#function_execution_timeline).

If your code looks like this:

```js
const {CloudTasksClient} = require('@google-cloud/task');
// CloudTasksClient, could be Storage, or Vision, or Speech.
const task = new CloudTasksClient();
export.helloWorld = (req, res) => {
  res.send('hello world');
}
```

Instead write:

```js
const {CloudTasksClient} = require('@google-cloud/task');
// CloudTasksClient, could be Storage, or Vision, or Speech.
let task;
export.helloWorld = async (req, res) => {
  if (!task) {
    task = new CloudTasksClient();
  }
  await task.doSomething(); // or storage.doSomething(), etc.
  res.send('hello world');
}
```

There is [work in motion to shield folks from this issue](https://github.com/googleapis/gapic-generator-typescript/issues/287). But, as described in the execution timeline, the core of the problem is that asynchronous work cannot happen outside of the request cycle of your function.

**Comment by vmal** [comment]
> @vmal @kthaas @omersadika @calclavia, we believe we've isolated this problem to creating a client in the global scope outside of your function [see: function execution lifetime](https://cloud.google.com/functions/docs/concepts/exec#function_execution_timeline).
> 
> If your code looks like this:
> 
> ```js
> const {CloudTasksClient} = require('@google-cloud/task');
> // CloudTasksClient, could be Storage, or Vision, or Speech.
> const task = new CloudTasksClient();
> export.helloWorld = (req, res) => {
>   res.send('hello world');
> }
> ```
> 
> Instead write:
> 
> ```js
> const {CloudTasksClient} = require('@google-cloud/task');
> // CloudTasksClient, could be Storage, or Vision, or Speech.
> let task;
> export.helloWorld = async (req, res) => {
>   if (!task) {
>     task = new CloudTasksClient();
>   }
>   await task.doSomething(); // or storage.doSomething(), etc.
>   res.send('hello world');
> }
> ```
> 
> There is [work in motion to shield folks from this issue](https://github.com/googleapis/gapic-generator-typescript/issues/287). But, as described in the execution timeline, the core of the problem is that asynchronous work cannot happen outside of the request cycle of your function.

Thank you, seems to have fixed our issue for now since we made sure that we avoid any global scopes for functions using google libarary.

**Comment by bcoe** [comment]
:wave: I'm locking this issue, because I believe we've isolated the underlying problem, which was instantiating (`new`ing) `@google-cloud/` client libraries in the global scope. 

This is because the libraries create asynchronous work, which is then immediately throttled, because it happens outside of a function's execution context. I've documented this behavior [here]([see: function execution lifetime](https://cloud.google.com/functions/docs/concepts/exec#function_execution_timeline) and [here](https://cloud.google.com/functions/docs/bestpractices/tips#performance).

To avoid this behavior, you simply need to modify your function to instantiate client libraries inside of your handler function, something like this:

```js
const {CloudTasksClient} = require('@google-cloud/task');
// CloudTasksClient, could be Storage, or Vision, or Speech.
const task = new CloudTasksClient();
export.helloWorld = (req, res) => {
  res.send('hello world');
}
```

Instead write:

```js
const {CloudTasksClient} = require('@google-cloud/task');
// CloudTasksClient, could be Storage, or Vision, or Speech.
let task;
export.helloWorld = async (req, res) => {
  if (!task) {
    task = new CloudTasksClient();
  }
  await task.doSomething(); // or storage.doSomething(), etc.
  res.send('hello world');
}
```

We also have work in progress [here](https://github.com/googleapis/gapic-generator-typescript/issues/287) to shield folks from the bug, which we're treating as a high priority.

If you think your issue is different than this one, please don't hesitate to open a new issue.

**Comment by bcoe** [comment]
I believe this issue has been addressed by:

1. the lazy initialization workaround (described above).
2. the work in https://github.com/googleapis/gapic-generator-typescript/issues/287, to attempt to make lazy initialization automatic (_this work is now available in all client libraries_).

**Comment by crwilcox** [comment]
Also adding, if you find that this is happening, you may consider using https://www.npmjs.com/package/is-online. Some serverless environments, kubernetes, will 'come up' before having connectivity and, since most apps start with auth, it will fail. Making sure you have outbound connectivity first can resolve this.
